frontend/i_types.py

```diff
@@ -0,0 +1,370 @@
+            else:
+                self.tree[t1] = [t2]
+
+    def __reachable(self, current, target, visited):
+        if current in visited or (not current in self.tree):
+            return False
+        if current == target:
+            return True
+
+        visited.add(current)
+        for adjacent in self.tree[current]:
+            if self.__reachable(adjacent, target, visited):
+                return True
+        return False
+
+    def reachable(self, source, target):
+        """Check if node2 is reachable from node1"""
+        return self.__reachable(source, target, set())
+
+
+subtypes_tree = SubtypesTree()
+
+class Type(metaclass=ABCMeta):
+    """This class is the top-parent of all possible inferred types of a Python program.
+    Every type has its own class that inherits this class.
+    """
+
+    @abstractmethod
+    def is_subtype(self, t):
+        """Check if this type is a subtype of the parameter type.
+
+        Arguments:
+            type (Type): The type to check the subtype relationship against.
+        """
+        pass
+
+
+    @abstractmethod
+    def get_name(self):
+        """Get the name of the type"""
+        pass
+
+    # Two types are considered identical if their names exactly match.
+    def __eq__(self, other):
+        return self.get_name() == other.get_name()
+
+    def __hash__(self):
+        return hash(self.get_name())
+
+    def __repr__(self):
+        return self.get_name()
+
+class TClass(Type):
+    """Type given to a class.
+
+    Attributes:
+        name (str): Class name.
+        context (Context): Contains a mapping between variables and types within the scope of this class.
+    """
+
+    def __init__(self, name, context=Context()):
+        self.name = name
+        self.context = context
+
+    def is_subtype(self, t):
+        if not isinstance(t, TClass):
+            return False
+        return subtypes_tree.reachable(self.name, t.name)
+
+class TNumber:
+    pass
+
+class TSequence:
+    pass
+
+class TMutableSequence(TSequence):
+    pass
+
+class TImmutableSequence(TSequence):
+    pass
+
+class TNone(Type):
+
+    def is_subtype(self, t):
+        return isinstance(t, TNone)
+
+    def get_name(self):
+        return "None"
+
+class TBool(TClass, TNumber):
+
+    def __init__(self):
+        self.strength = 0
+        self.name = "bool"
+
+    def get_name(self):
+        return self.name
+
+
+class TInt(TClass, TNumber):
+
+    def __init__(self):
+        self.strength = 1
+        self.name = "int"
+
+    def get_name(self):
+        return self.name
+
+
+class TFloat(TClass, TNumber):
+
+    def __init__(self):
+        self.strength = 2
+        self.name = "float"
+
+    def get_name(self):
+        return self.name
+
+class TComplex(TClass, TNumber):
+
+    def __init__(self):
+        self.strength = 3
+        self.name = "complex"
+
+    def get_name(self):
+        return self.name
+
+
+class TString(TClass, TImmutableSequence):
```

Comment:
can this (and similarly for the other sequences) just be a subtype of TImmutableSequence?